mount/mount_windows.go

```diff
@@ -46,22 +72,42 @@ func (m *Mount) mount(target string) error {
 	if err = hcsshim.ActivateLayer(di, layerID); err != nil {
 		return fmt.Errorf("failed to activate layer %s: %w", m.Source, err)
 	}
+	defer func() {
+		if err != nil {
+			hcsshim.DeactivateLayer(di, layerID)
+		}
+	}()
 
 	if err = hcsshim.PrepareLayer(di, layerID, parentLayerPaths); err != nil {
 		return fmt.Errorf("failed to prepare layer %s: %w", m.Source, err)
 	}
+	defer func() {
+		if err != nil {
+			hcsshim.UnprepareLayer(di, layerID)
+		}
+	}()
 
-	// We can link the layer mount path to the given target. It is an UNC path, and it needs
-	// a trailing backslash.
-	mountPath, err := hcsshim.GetLayerMountPath(di, layerID)
+	volume, err := hcsshim.GetLayerMountPath(di, layerID)
 	if err != nil {
-		return fmt.Errorf("failed to get layer mount path for %s: %w", m.Source, err)
+		return fmt.Errorf("failed to get volume path for layer %s: %w", m.Source, err)
 	}
-	mountPath = mountPath + `\`
 
-	if err = os.Symlink(mountPath, target); err != nil {
-		return fmt.Errorf("failed to link mount to target %s: %w", target, err)
+	if err = bindfilter.ApplyFileBinding(target, volume, readOnly); err != nil {
+		return fmt.Errorf("failed to set volume mount path for layer %s: %w", m.Source, err)
 	}
+	defer func() {
+		if err != nil {
+			bindfilter.RemoveFileBinding(target)
+		}
+	}()
+
+	// Add an Alternate Data Stream to record the layer source.
+	// See https://docs.microsoft.com/en-au/archive/blogs/askcore/alternate-data-streams-in-ntfs
+	// for details on Alternate Data Streams.
+	if err = os.WriteFile(filepath.Clean(target)+":"+sourceStreamName, []byte(m.Source), 0666); err != nil {
+		return fmt.Errorf("failed to record source for layer %s: %w", m.Source, err)
+	}
```

Comment:
Clever! @TBBle

Reply:
Although, kind of annoying we even need to do this

Reply:
This wasn't my idea, I cadged it from the even older "Mount Windows" PR #2287.

It is a shame that there's no way to get the _layer_ being unmounted from either containerd (Unmount only takes the mount point name, because that's how the umount syscall works on Linux...) or from the OS (because HCS's XXXLayer API lacks a way to lookup layer details from the WCIFS virtual volume; I don't recall if the computestorage API changed this).

Not to say this isn't a reasonable solution. It even has the nice effect that deleting the mount point folder automatically removes the ADS data too so there's no out-of-sync data feasible, and makes the relevant data available to some kind of inspection/recovery tool if needed, e.g. to unblock layer deactivation in the case of containerd death.

Hopefully we never hit a problem with this from some tool that goes weird in the face of ADS data, like Go's filepath.walk does in the face of junction points.